src/sectmap.cpp

```diff
@@ -0,0 +1,198 @@
+/*
+ *
+ * sector map class for DeadDiscReader
+ * Copyright (C) 2006-2007,2009, Gennady "ShultZ" Kozlov <qpxtool@mail.ru>
+ * it uses QPxTool SCSI transport library
+ *
+ */
+
+#define _FILE_OFFSET_BITS 64
+
+#include <cstdio>
+#include <cstdlib>
+#include <sys/stat.h>
+
+#include "sectmap.hpp"
+
+long fsize(FILE *f) {
+  struct stat st;
+  fstat(fileno(f), &st);
+  return st.st_size;
+}
+
+smap::smap(char *fn, unsigned int sects) {
+  sectors = sects;
+  fname = fn;
+  blocks = (sectors / map_block_sz) + !!(sectors % map_block_sz);
+  arr = (map_block *)malloc(sizeof(map_block) * blocks);
+  fill(BM_WAIT);
+  mutex = new Mutex();
+  printf("* map: created for %d sectors\n", sectors);
+}
+
+smap::~smap() {
+  save();
+  delete mutex;
+  delete arr;
+}
+
+mape smap::get(unsigned int sector) {
+  if (sector > sectors)
+    return BM_INV;
+  unsigned int row = sector / map_block_sz;
+  unsigned int col = sector % map_block_sz;
+  return arr[row][col];
+}
+
+int smap::is_done() { return (sectors - get_done()); }
+
+int smap::get_wait() {
+  unsigned int cnt = 0;
+  for (unsigned int i = 0; i < sectors; i++)
+    if (!get(i))
+      cnt++;
+  return cnt;
+}
+
+int smap::get_read() {
+  unsigned int cnt = 0;
+  for (unsigned int i = 0; i < sectors; i++)
+    if (get(i) == BM_READ)
+      cnt++;
+  return cnt;
+}
+
+int smap::get_done() {
+  unsigned int cnt = 0;
+  for (unsigned int i = 0; i < sectors; i++)
+    if (get(i) == BM_DONE)
+      cnt++;
+  return cnt;
+}
+
+int smap::get_fail() {
+  unsigned int cnt = 0;
+  for (unsigned int i = 0; i < sectors; i++)
+    if (get(i) == BM_FAIL)
+      cnt++;
+  return cnt;
+}
+
+int smap::get_tot() { return sectors; }
+
+unsigned int smap::get_next(unsigned int *lba, mape state, int *count) {
+  unsigned int offs = 0;
+  int cnt = 0;
```

Comment:
Codacy found an issue: [The scope of the variable 'cnt' can be reduced.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)